### DISCUSSION TITLE
TLS and CNs - golang 1.15

### DISCUSSION THREAD

**Discussion by vjanelle** [choria-io/general#5] [General]
Golang 1.15 depreciated CAs which are created without SANs - this is something that was depreciated 20 years ago.

Currently we have (at least) one test CA in our test suite which was created this way - IIRC this was because the CA could've been made with puppet?

Is this behavior in the current versions of puppet?  If so, when was that changed?

**Comment by vjanelle** [comment]
Grabbing the latest puppetserver image, looks like it doesn't create CAs with a SAN >.<

**Comment by ripienaar** [comment]
Yup :( and even if recent ones did we would have like years more of this dumb things as most old certs are this way and certs can be valid for 5 years

little choice but to write our own validation code that gets called only for certs without them

**Comment by vjanelle** [comment]
Yup - the "ignoreCN" variable is created at runtime, so unfortunately the only thing we could do is to ensure the variable is set before we start the process.

**Comment by ripienaar** [comment]
1.16 removes that variable. Will need to code own validation

**Reply by vjanelle** [reply]
Hmm, when I looked at the source it was still there.. https://github.com/golang/go/blob/master/src/crypto/x509/verify.go#L22

**Comment by ripienaar** [comment]
Not sure where I saw it’s targeted, anyway we can’t expect it to be there long nor can we ensure it’s set for CLI interactions :(

**Comment by vjanelle** [comment]
https://github.com/choria-io/go-choria/pull/1057 

Here's a potential hack - if a SAN doesn't exist, create one and shove the common name into it.  Should probably be hidden behind a configuration flag (that defaults to on..)

**Comment by ripienaar** [comment]
If it works I am all for it :)

**Comment by marc-brou** [comment]
Hey guys,
I have been trying to setup the latest version of Choria in a new environment and I seem to be hitting a lot of Cert missing SAN issues, I have seen [Issue#1060](https://github.com/choria-io/go-choria/issues/1060) and tried setting plugin.security.support_legacy_certificates to true, but it didn't seem to have any effects, altho I did not find how to set this with Hiera so I added manually to all the conf files under /etc/choria and also added a plugin.d/security.cfg file with support_legacy_certificates = true inside.   Would appreciate info on correct way to add this setting so it doesn't get removed by puppet.

I can get nodes to connect to the network broker fine, however the federated broker gives this error:
{"cluster":"phxl","component":"federation","connection":"choria_nats_ingest","level":"warning","msg":"Initial connection to the Broker failed on try 22790: x509: certificate relies on legacy Common Name field, use SANs or temporarily enable Common Name matching with GODEBUG=x509ignoreCN=0","time":"2021-03-11T03:45:27-05:00","worker":"choria_nats_ingest","worker_instance":2}

and client seems to have issues too, for example:
$ choria ping
WARN[0000] Initial connection to the Broker failed on try 1: x509: certificate is not valid for any names, but wanted to match   component=client connection=myuser.mcollective-9745691595a94dffaac2783022172293-receiver2

I am no Certificate Expert but you are talking about the CA cert not having a SAN, which seems weird to me, I understand for normal certs, but why would a root CA need a SAN? would my intermediate CA that puppet uses having a SAN fix the issues?

**Comment by ripienaar** [comment]
Hmm. I have to limit this code only to the clients but I guess I have a bug where I don’t handle the federation clients. 

I agree it’s weird and unnecessary but this is a change in tbe Go language that cannot be opted out and that is unbelievably stupid. So we are just trying to work around this idiocy as best we can :(

**Comment by ripienaar** [comment]
@marc-brou you are using Puppet CA? its weird you get this for normal client too

**Reply by marc-brou** [reply]
Yes, using Puppet CA for choria, our puppet CA is an intermediate CA signed by our internal root CA, neither have SANs.

All my servers can connect to the choria broker just fine, federation connection doesn't work, and choria client (setup using choria enroll) on the same server as the broker fails when I test with choria ping with the error in the post above.  (Note that broker is not running on the same server as puppetserver)

**Reply by ripienaar** [reply]
If you run that choria ping with —debug so you see the lines saying it’s using legacy tls?

**Reply by marc-brou** [reply]
With debug the only extra I get seems to be some failed SRV lookups, which seems weird since I disabled the federation as it is not working, and it's not using the domain I set in hiera's choria::srvdomain:, uses the server's domain instead.
```
$ choria --debug ping
DEBU[0000] Logging at debug level due to CLI override   
DEBU[0000] Starting 3 receivers on mcollective.reply.926d084241fc491acce94ff7de5a1aaa.f3a676a5606e4b2ebf220a2a50c304ca  component=client
DEBU[0000] Attempting SRV lookup for _mcollective-federation_server._tcp.mydomain.com
DEBU[0000] Attempting SRV lookup for _mcollective-federation_server._tcp.mydomain.com
DEBU[0000] Failed to resolve _mcollective-federation_server._tcp.mydomain.com: lookup _mcollective-federation_server._tcp.mydomain.com on 10.152.56.101:53: no such host 
DEBU[0000] Attempting SRV lookup for _x-puppet-mcollective_federation._tcp.mydomain.com 
DEBU[0000] Failed to resolve _mcollective-federation_server._tcp.mydomain.com: lookup _mcollective-federation_server._tcp.mydomain.com on 10.152.56.101:53: no such host 
DEBU[0000] Attempting SRV lookup for _x-puppet-mcollective_federation._tcp.mydomain.com 
DEBU[0000] Failed to resolve _x-puppet-mcollective_federation._tcp.mydomain.com: lookup _x-puppet-mcollective_federation._tcp.mydomain.com on 10.152.56.101:53: no such host 
INFO[0000] Attempting to connect to:                     component=client connection=choriasvc.mcollective-7d50e14e5d714e688f4f0d4eb0ae79c6-receiver0
DEBU[0000] Failed to resolve _x-puppet-mcollective_federation._tcp.mydomain.com: lookup _x-puppet-mcollective_federation._tcp.mydomain.com on 10.152.56.101:53: no such host 
INFO[0000] Attempting to connect to:                     component=client connection=choriasvc.mcollective-7d50e14e5d714e688f4f0d4eb0ae79c6-receiver1
WARN[0000] Initial connection to the Broker failed on try 1: x509: certificate is not valid for any names, but wanted to match   component=client connection=choriasvc.mcollective-7d50e14e5d714e688f4f0d4eb0ae79c6-receiver1
[....]
```
I manually added "support_legacy_certificates = true" in /etc/choria/client.conf, didn't seem to change anything.

**Reply by ripienaar** [reply]
This is a different error: “ Initial connection to the Broker failed on try 1: x509: certificate is not valid for any names” it is not related to the legacy tls setup. 

Please show your configs and give us detail of how you create certs. Above indicates the cert on your broker is really weird

**Reply by ripienaar** [reply]
The setting is also not right (but wo t fix this problem)

https://github.com/choria-io/go-choria/blob/main/CONFIGURATION.md#pluginsecuritysupport_legacy_certificates

**Reply by marc-brou** [reply]
I'd assume the broker is just using the puppet agent's ssl certs, and the client cert is signed by the puppet server and created with "choria enroll".    I have just deleted the puppet agent's ssl folder and the client's user ~/.puppetlabs ssl folder and recreated them and still get same error.
my configs seems pretty standard:
[choria_configs.txt](https://github.com/choria-io/general/files/6151688/choria_configs.txt)
let me know if you want to see other config files.

**Reply by ripienaar** [reply]
Fairly standard stuff. Can you show with OpenSSL command the cert the broker is using? 

It should log it at debug. Else try “sudo mco choria show_config —config broker.conf” and show the cert

the problem here appears to be that the cert has no Common Name either. 

For sure federation isn’t going to work today. I need to rejig a bunch of things so probably only next release.

**Reply by marc-brou** [reply]
here is the /etc/puppetlabs/puppet/ssl/certs on the broker (choria) server: (changed my actual domain for mydomain.com because internet;)
```
[root@phxl1-choria-001 certs]# openssl x509 -text -in phxl1-choria-001.mydomain.com.pem
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 340 (0x154)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=US, ST=California, L=My City, O=My Corporation, OU=My Service Operations, DC=l, DC=puppet, CN=Lab Puppet Signing Certificate
        Validity
            Not Before: Mar 15 15:44:18 2021 GMT
            Not After : Mar 14 15:44:18 2031 GMT
        Subject: CN=phxl1-choria-001.mydomain.com
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (4096 bit)
                Modulus:
                    [snip...]
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            Netscape Comment: 
                Puppet Server Internal Certificate
            X509v3 Authority Key Identifier: 
                keyid:E5:D0:1A:0A:D6:DC:15:9D:49:C4:7B:CA:0D:0F:17:55:41:05:1A:A8

            X509v3 Subject Key Identifier: 
                D5:66:0F:EF:6D:47:75:FA:6C:97:0A:84:63:D5:21:1E:68:69:DA:EE
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Extended Key Usage: critical
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
    Signature Algorithm: sha256WithRSAEncryption
[...]
```
And the output of your command does say Valid SSL setup: no, but I did run enroll and it says it's already been run, pasted below:
```
[choriasvc@phxl1-choria-001 choria]$ sudo mco choria show_config --config broker.conf
Active Choria configuration:

The active configuration used in Choria comes from using Puppet AIO defaults, querying SRV
records and reading configuration files.  The below information shows the completely resolved
configuration that will be used when running MCollective commands

MCollective related:

    MCollective Version: 2.24.1
         Choria Version: 0.19.0
     Client Config File: /etc/choria/client.conf
     Active Config File: broker.conf
      Plugin Config Dir: ./plugin.d
      Using SRV Records: true
              Federated: false
             SRV Domain: choria
               NATS NGS: false
     Middleware Servers: phxl1-choria-001.mydomain.com:4222

Puppet related:

       Puppet Server: puppet:8140
     PuppetCA Server: puppet:8140
     PuppetDB Server: puppet:8081
     Discovery Proxy: not using a proxy
      Facter Command: /opt/puppetlabs/bin/facter
       Facter Domain: mydomain.com

Security setup:

     Valid SSL Setup: no try running 'choria enroll'
   Security Provider: puppet
            Certname: phxl1-choria-001.mydomain.com
       SSL Directory: /etc/puppetlabs/puppet/ssl (found)
  Client Public Cert: /etc/puppetlabs/puppet/ssl/certs/phxl1-choria-001.mydomain.com.pem (found)
  Client Private Key: /etc/puppetlabs/puppet/ssl/private_keys/phxl1-choria-001.mydomain.com.pem (found)
             CA Path: /etc/puppetlabs/puppet/ssl/certs/ca.pem (found)
            CSR Path: /etc/puppetlabs/puppet/ssl/certificate_requests/phxl1-choria-001.mydomain.com.pem (absent)
      Public Cert CN: phxl1-choria-001.mydomain.com (match)

Active Choria configuration settings as found in configuration files:

              choria.srv_domain: choria
           choria.stats_address: ::
              choria.stats_port: 8222
          choria.broker_network: true
  choria.network.listen_address: ::
     choria.network.client_port: 4222
       choria.network.peer_port: 4223
       choria.broker_federation: true
      choria.federation.cluster: phxl
     choria.federation.instance: phxl1-choria-001.mydomain.com
        choria.middleware_hosts: phxl1-choria-001.mydomain.com:4222

[choriasvc@phxl1-choria-001 choria]$ choria enroll
Enrolling with the Security System using certname choriasvc.mcollective
choria: error: Could not enroll: already have all files needed for SSL operations
[choriasvc@phxl1-choria-001 choria]$ 
```

**Reply by ripienaar** [reply]
In the log from choria --debug ping - there are gaps, did you just entirely remove names from there rather than replace them with something? its kind of important, please replace domain with mydomain.com consistently so its possible to see whats happening.

**Reply by marc-brou** [reply]
I did not remove anything, just replaced actual domain by mydomain.com, now that you mention it, it does seem strange, but there is actually a long blank in the logs at:  Attempting to connect to:                     component=client, as if the target was blank, which might be because it's not able to resolve the srv record just above, which I would expect to try to resolve as _mcollective-federation_server._tcp.choria.mydomain.com as I set my srvdomain=choria, then I expected it to add search domain from resolv.conf but it seems to take my server's domain fact instead.

**Reply by ripienaar** [reply]
OK, I have no idea whats going on - I'd start simple, disable SRV and configure manually and go from there - get basics working right.

We did recently make some SRV changes, could well have broken something.

Still, federation broker wont work with this release.

**Reply by marc-brou** [reply]
ok, will do :)
Thanks for the help.

**Reply by marc-brou** [reply]
After adding the manual config section to hiera as per https://choria.io/docs/deployment/dns/ I noticed that the client.conf was not modified by the puppet run, started to wonder where the client was taking it's connection info as the broker.conf and server.conf are both root owned and not readable by my user.   I tried manually adding "plugin.choria.srv_domain = choria" to client.conf and the client is now trying to resolve the correct SRV and can connect to an actual target, not blank anymore, so you were right that it had nothing to do with legacy certs, altho not sure how I will get puppet to not overwrite my client.conf :(
It also looks like my nodes are not getting discovered as while the ping command works, I do not get any replies, but that's probably another wrong config somewhere.

**Reply by ripienaar** [reply]
You can set these values by puppet too. 

For the nodes not responding. See their logs.

**Reply by marc-brou** [reply]
I am already setting choria::srvdomain:  in hiera, is there a different setting affecting the client.conf ?

**Reply by marc-brou** [reply]
ok, got it to add srv to client.conf by adding in hiera:
```
mcollective::client_config:
    plugin.choria.srv_domain: choria
```
and commenting out: #plugin.choria.federation.collectives from the client.conf that was probably there from my test with the federation stuff which seems was causing the nodes to not reply.
Seems to all be working fine without the federation enabled now :)
Thanks again for the help!

**Reply by ripienaar** [reply]
cool, srvdomain vs srv_domain and setting this once only is high on the list of things to fix as we work to simplify the modules

glad its sorted now

**Reply by ripienaar** [reply]
Tomorrows nightly build will probably also fix federation if you wanted to test things

**Reply by marc-brou** [reply]
Yup, Federation seems to work fine with the nightly build.  Thanks again!
